### PR TITLE
WIP: IPAM plugin after taking care of the comments

### PR DESCRIPTION
**Pull request by leodotcloud**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,6 @@
 /.dapper
 /bin
 *.swp
+*.diff
```
Comment (alena1108) [inline .gitignore:4]: what is this file?

File: main.go [hunk 2]
```diff
@@ -1,9 +1,62 @@
...
+			log.SetOutput(f)
+			defer f.Close()
+		}
+	}
+
+	log.Debugf("rancher-cni-ipam: cmdAdd: invoked")
+	log.Debugf("rancher-cni-ipam: %s", fmt.Sprintf("args: %#v", args))
+	log.Debugf("rancher-cni-ipam: %s", fmt.Sprintf("ipamConf: %#v", ipamConf))
+
+	ipf, err := metadata.NewIPFinderFromMetadata()
```
Comment (alena1108) [inline main.go:37]: is it possible to initiate the finderFromMetadata just once? not on every request. We can re-use the same client

File: main.go [hunk 3]
```diff
@@ -1,9 +1,62 @@
...
+	ipString := ipf.GetIP(args.ContainerID)
+
+	log.Debugf("rancher-cni-ipam: %s", fmt.Sprintf("ip: %#v", ipString))
+
+	ip, ipnet, err := net.ParseCIDR(ipString + "/16")
+	if err != nil {
+		return err
+	}
+
+	// TODO: if ip is NULL, return err
```
Comment (alena1108) [inline main.go]: are we going to address this TODO?

File: main.go [hunk 4]
```diff
@@ -1,9 +1,62 @@
...
+
+	ipf, err := metadata.NewIPFinderFromMetadata()
+	if err != nil {
+		return err
+	}
+	ipString := ipf.GetIP(args.ContainerID)
+
+	log.Debugf("rancher-cni-ipam: %s", fmt.Sprintf("ip: %#v", ipString))
+
+	ip, ipnet, err := net.ParseCIDR(ipString + "/16")
```
Comment (alena1108) [inline main.go]: is netmask we are checking against, configurable? as well as cidr ip range?

File: ipfinder/metadata/ip_finder_metadata.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+		return nil, err
+	}
+	return &IPFinderFromMetadata{m}, nil
+}
+
+// GetIP returns the IP address for the given container id, return an empty string
+// if not found
+func (ipf *IPFinderFromMetadata) GetIP(cid string) string {
+
+	for i := 0; i < 600; i++ {
```
Comment (alena1108) [inline ipfinder/metadata/ip_finder_metadata.go]: the wait 600x500miliseconds  looks a bit long to me

File: ipfinder/metadata/ip_finder_metadata.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+		containers, err := ipf.m.GetContainers()
+		if err != nil {
+			log.Errorf("rancher-cni-ipam: Error getting metadata containers: %v", err)
+			return empty
+		}
+
+		for _, container := range containers {
+			if container.ExternalId == cid {
+				log.Infof("rancher-cni-ipam: got ip: %v", container.PrimaryIp)
```
Comment (alena1108) [inline ipfinder/metadata/ip_finder_metadata.go:43]: shall we log this in the caller stack? so we log the information about the ip address returned by ip-finder in generic way

File: ipfinder/metadata/ip_finder_metadata.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+				return container.PrimaryIp
```
Comment (alena1108) [inline ipfinder/metadata/ip_finder_metadata.go:44]: one comment on this note: how do we deal with the host level networking? does k8s talk to cni plugin to obtain that info as well? if so, we should make sure we cover this use case as well.

File: main.go [hunk 8]
```diff
@@ -1,9 +1,62 @@
...
+	log.Debugf("rancher-cni-ipam: cmdAdd: invoked")
+	log.Debugf("rancher-cni-ipam: %s", fmt.Sprintf("args: %#v", args))
+	log.Debugf("rancher-cni-ipam: %s", fmt.Sprintf("ipamConf: %#v", ipamConf))
+
+	ipf, err := metadata.NewIPFinderFromMetadata()
+	if err != nil {
+		return err
+	}
+	ipString := ipf.GetIP(args.ContainerID)
+
```
Comment (alena1108) [inline main.go:45]: we should raise an error if the ip address returned by GetIp, is empty, right away w/o future processing

File: ipfinder/metadata/ip_finder_metadata.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+import (
+	log "github.com/Sirupsen/logrus"
+	"github.com/rancher/go-rancher-metadata/metadata"
+	"time"
+)
+
+const (
+	metadataURL = "http://rancher-metadata/latest"
+	empty       = ""
```
Comment (alena1108) [inline ipfinder/metadata/ip_finder_metadata.go]: we should rename this param so it's use is obvious. It is a bit an overkill though to create a constant for an empty string :)

File: ipfinder/metadata/ip_finder_metadata.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
+package metadata
+
+import (
+	log "github.com/Sirupsen/logrus"
+	"github.com/rancher/go-rancher-metadata/metadata"
+	"time"
+)
+
+const (
+	metadataURL = "http://rancher-metadata/latest"
```
Comment (alena1108) [inline ipfinder/metadata/ip_finder_metadata.go]: please point to the specific version of metadata. Using "latest" can bite you later

File: main.go (hunk 2 above)
Comment (leodotcloud) [inline main.go:37]: CNI IPAM is an executable which gets invoked, so it will be done for every request.

File: .gitignore (hunk 1 above)
Comment (leodotcloud) [inline .gitignore:4]: Usually git diff are stored in temporary files using the prefix .diff

File: ipfinder/metadata/ip_finder_metadata.go (hunk 9 above)
Comment (leodotcloud) [inline ipfinder/metadata/ip_finder_metadata.go]: Ok, will change it :)

File: ipfinder/metadata/ip_finder_metadata.go (hunk 10 above)
Comment (leodotcloud) [inline ipfinder/metadata/ip_finder_metadata.go]: Ok.

File: ipfinder/metadata/ip_finder_metadata.go (hunk 5 above)
Comment (leodotcloud) [inline ipfinder/metadata/ip_finder_metadata.go]: @ibuildthecloud what do you suggest?

File: ipfinder/metadata/ip_finder_metadata.go (hunk 6 above)
Comment (leodotcloud) [inline ipfinder/metadata/ip_finder_metadata.go:43]: It's logged there as well with more information.

```
log.Infof("rancher-cni-ipam: %s", fmt.Sprintf("r: %#v", r))
return r.Print()
```

File: ipfinder/metadata/ip_finder_metadata.go (hunk 7 above)
Comment (leodotcloud) [inline ipfinder/metadata/ip_finder_metadata.go:44]: Once we specify the CNI plugin while launching kubelet, there is no more host level IP address like docker. Everything goes through the CNI plugin.

File: main.go (hunk 8 above)
Comment (leodotcloud) [inline main.go:45]: agree

File: main.go (hunk 3 above)
Comment (leodotcloud) [inline main.go]: Addressed.

File: main.go (hunk 4 above)
Comment (leodotcloud) [inline main.go]: Only the mask is configurable, right now we are getting the IP directly from cattle. There is no concept of separate subnet range for different hosts. I have the changes for the mask in the configuration. Will submit them in the next PR.

File: ipfinder/metadata/ip_finder_metadata.go (hunk 5 above)
Comment (ibuildthecloud) [inline ipfinder/metadata/ip_finder_metadata.go]: @leodotcloud 2 minutes

File: ipfinder/metadata/ip_finder_metadata.go (hunk 7 above)
Comment (alena1108) [inline ipfinder/metadata/ip_finder_metadata.go:44]: What will happen is user specifies "hostNetwork" in their pod template spec? He won't be able to launch it? @ibuildthecloud

Comment (leodotcloud) [inline ipfinder/metadata/ip_finder_metadata.go:44]: @alena1108 I just launched a pod spec with hostNetwork: true and k8s doesn't talk to the CNI plugin for this. Thinking about it, I felt natural as there is on need to communicate with the CNI plugin for host level networking.

**Pull request merged**